README.md
# Minimal WebDriver

[![Build Status]](https://travis-ci.org/mantoni/min-webdriver)
[![SemVer]](http://semver.org)
[![License]](https://github.com/mantoni/min-webdriver/blob/master/LICENSE)

Pipe scripts to browsers using the Selenium [WebDriver protocol][].

- [SauceLabs][] support
- [Appium][] support for mobile testing
- Selenium Server 2 support
- Concurrent test runs
- No web server required

## Install

```
npm install min-wd
```

## Usage

Put a config file name `.min-wd` in your project directory:

```
{
  "hostname": "localhost",
  "port": 4444,
  "browsers": [{
    "name": "internet explorer",
    "version": "10"
  }, {
    "name": "chrome"
  }]
}
```

You can also have the `.min-wd` file be loaded as a module:

```
var hostname = true ? "localhost" : "otherhost";

module.exports = {
  "hostname": hostname,
  "port": 4444,
  "browsers": [{
    "name": "internet explorer",
    "version": "10"
  }, {
    "name": "chrome"
  }]
}
```


Alternatively, add a `webdriver` property with the configs to your
`package.json` file.

Assuming `my-script.js` contains this:

```js
console.log('Hello %s!', 'browser');
process.exit(0);
```

Use with [browserify][]:

```
$ browserify -p min-wd my-script.js
# internet explorer 9:
Hello browser!
# chrome *:
Hello browser!
```

Additional Selenium capabilities and browser-specific capabilities can be
specified with the `capabilities` property:

```
{
  "hostname": "localhost",
  "port": 4444,
  "browsers": [{
    "name": "chrome",
    "capabilities": {
      "chromeOptions": {
        "args": ["--headless", "--disable-gpu"]
      }
    }
  }]
}
```

## Shareable configuration package

Inspired by [ESLint's shareable configuration package][Eslint Shareable Config],
there is the possibility of storing selenium and browser-specific capabilities
in a separate npm package. This way one config can be shared amongst many npm 
packages instead of them each having their own `.min-wd` file.

### Usage

Create a npm package and name it using the prefix `min-wd-config-` 
(eg. min-wd-config-myapp).

The shareable config must contain a js file that exports
an object with the configuration or a `.min-wd` file. 

Example: index.js
```
module.exports = {
  "hostname": "localhost",
  "port": 4444,
  "browsers": [{
    "name": "chrome"
  }]
}
```

In the shareable config's package.json you will need to reference the file 
containing the config:

```
{
  "name": "min-wd-config-myapp",
  "version": "^1.0.0",
  "webdriver": {
    "extends": "./index.js"
  }
}
```

In the shareable config's dependent, that is the package that consumes the 
shareable config, the config name without the `min-wd-config-` prefix needs to 
be specified:

```
{
  "name": "myapp",
  "webdriver": {
    "extends": "myapp"
  },
  "devDependencies": {
    "min-wd-config-myapp": "^1.0.0"
  }
}
```

When using a shareable config and a `.min-wd` file, the latter will take 
precedence.


## SauceLabs

Export your SauceLabs credentials:

```
export SAUCE_USERNAME="your-user-name"
export SAUCE_ACCESS_KEY="your-access-key"
```

Enable SauceLabs in your `.min-wd` file:

```
{
  "sauceLabs": true,
  "browsers": [...]
}
```

See ["Supported options"](#supported-options) for additional SauceLabs specific
options and ["SauceLabs on Travis"](#saucelabs-on-travis) on how to run
min-webdriver tests on Travis.

## Appium

Note: This has only be tested on Mac OS X High Sierra with the iOS Simulator so
far. If you have successfully tested with other configurations, please file an
issue so that we can extend the docs.

Setup for iOS Simulator on Mac OS X:

- `npm install -g appium`
- `brew install carthage`
- Configure your `.min-wd` file like this:

```
{
  "hostname": "localhost",
  "port": 4723,
  "browsers": [{
    "name": "Safari",
    "platformName": "iOS",
    "platformVersion": "11.2",
    "deviceName": "iPhone Simulator"
  }]
}
```

- Run `appium` which should start a server on port 4723
- Run your tests

## BrowserStack

Export your BrowserStack credentials:

```
export BROWSERSTACK_USERNAME="your-user-name"
export BROWSERSTACK_ACCESS_KEY="your-access-key"
```

Example `.min-wd` file:

```
module.exports = {
  "hostname": "hub-cloud.browserstack.com",
  "port": 80,
  "browsers": [{
    "name": "chrome",
    "capabilities": {
      "browser": "Chrome",
      "browserstack.user": process.env.BROWSERSTACK_USERNAME,
      "browserstack.key": process.env.BROWSERSTACK_ACCESS_KEY
    }
  }]
}
```

## Loading a web page

By default, min-webdriver will fork a new browser and inject the given script
straight away without loading any web page. If you want to run your test cases
in the context of a web page, you can configure the start page in the `.min-wd`
file:

```
{
  "url": "http://my-test-page"
}
```

## Mocha Support

Testing with [Mocha][] requires [mocaccino][]:

```
$ browserify -p mocaccino -p min-wd my-test.js
```

If this is your use case, make sure to give [Mochify][] a try.

## Timeouts

The default timeout for the log polling script is 10 seconds. If you have long
running test cases that don't print anything for more than 10 seconds, you can
increase the timeout by adding a `timeout` property to your config:

```
"timeout": 20000
```

**Notice:** This option is not used if explicitly setting the `asyncPolling`
option to `false`.

## API

Use min-wd programatically with browserify like this:

```
var browserify = require('browserify');
var minWd = require('min-wd');

var b = browserify();
b.plugin(minWd, { timeout : 0 });
```

### Supported options

- `wdFile` specify the location of the `.min-wd` config file. Defaults to `.min-wd`.
- `sauceLabs` whether to run tests with `saucelabs`. Defaults to `false`.
- `hostname` the host to connect to. Defaults to `localhost`. If `sauceLabs` is
  `true`, `ondemand.saucelabs.com` is used.
- `port` the port to connect to. Defaults to `4444`. If `sauceLabs` is `true`,
  `80` is used.
- `asyncPolling` whether to use async polling when looking for test results.
  Defaults to `true`.
- `timeout` if a script does not respond to log polling calls for this amount
  of milliseconds, the test run is aborted. Defaults to 10 seconds.
- `url` the URL to open in each browser. Defaults to no URL.
- `closeOnSuccess` whether to close browsers on success. Defaults to `true`.
- `closeOnError` whether to close browsers on error. Defaults to `true`.
- `browsers` an array of browser config objects, each with these properties:
    - `name` the name of the browser to launch, e.g. `chrome`, `firefox` or
      `internet explorer`
    - `version` the browser version to launch. Use `*` for any.
    - `url` an optional URL to launch for this browser
    - `capabilities` additional Selenium capabilities and browser-specific
      capabilities

Some options are only considered depending on the `asyncPolling` value:

  - `pollingInterval` sets the time interval between test log checks. Only
    apply if `asyncPolling` is `false`. Defaults to 1000 milliseconds.
  - `timeout` option won't apply if `asyncPolling` is set to `false` because
    the test log is checked manually respecting `pollingInterval`.

SauceLabs specific options that only apply if `sauceLabs` is set to `true`:

- `sauceJobName` sets a custom name for the SauceLabs job. If a `package.json`
  file exists in the current directory, this defaults to the package name.
- `BUILD_VAR` sets the environment variable name that contains a build number
  to set, e.g. `TRAVIS_BUILD_NUMBER`.

### SauceLabs on Travis

Assuming you have your linter and headless tests in an npm "test" script and
the SauceLabs tests in a script called "wd", then the SauceLabs builds can be
used by configuring your `.travis.yml` like this:

```
language: node_js
node_js:
  - "0.12"
  - "4.2"

env:
  global:
    - SAUCE_USERNAME=mantoni
    - secure: "your-secured-access-key"

script:
  - 'npm test'
  - 'if [ "x$TRAVIS_NODE_VERSION" = "x4.2" ]; then npm run wd; fi'
```

This will run `npm test` on Node 0.12 and 4.2 while running the "wd" script
only in the node 4.2 build.  See the Travis [documentation for encryption
keys][TravisKeys] for details about the `secure` config.

## Known issues and solutions

`min-webdriver` injects your script directly into the default page launched by
the Selenium driver. In some cases browsers behave strangely in this context.
Work around this by specifying a URL to a simple web page that is loaded before
running the tests:

```
{
  "browsers": [{
    "name": "Internet Explorer",
    "version": "9",
    "url": "http://my-server/doctype.html"
  }]
}
```

With this content in the `doctype.html`:

```html
<!DOCTYPE html><html><head><meta encoding="utf-8"></head><body></body></html>
```

You can also specify a `"url"` for all browser on the root level.

Loading a page before injecting the scripts is solving these issues:

- IE 9 reporting it can't find `JSON` because the Selenium default page makes
  IE switch to quirks mode
- Error: `SECURITY_ERR: DOM Exception 18` because setting cookies is not
  allowed for `file://` URLs
- Error: `access to the Indexed Database API is denied in this context`
- localStorage being inaccessible.
- iOS Simulator does not report anything: Set `asyncPolling: false`

#### Usage with Microsoft Edge browser

To configure edge use `{ "name": "microsoftedge" }`. For the time being, MS
Edge doesn't support `asyncPolling` set to `true`. If you want to test with
that browser you must set `asyncPolling` to `false`.

## Compatibility

- Node 0.10, 0.12, 4.x, 5.x
- Selenium Server 2.39 or later

## License

MIT

[Build Status]: http://img.shields.io/travis/mantoni/min-webdriver.svg
[SemVer]: http://img.shields.io/:semver-%E2%9C%93-brightgreen.svg
[License]: http://img.shields.io/npm/l/min-wd.svg
[WebDriver protocol]: https://code.google.com/p/selenium/wiki/JsonWireProtocol
[browserify]: http://browserify.org
[Mocha]: http://mochajs.org/
[Mochify]: https://github.com/mantoni/mochify.js
[mocaccino]: https://github.com/mantoni/mocaccino.js
[TravisKeys]: https://docs.travis-ci.com/user/encryption-keys/
[SauceLabs]: https://saucelabs.com
[Appium]: http://appium.io
[Eslint Shareable Config]: https://eslint.org/docs/user-guide/configuring#using-a-shareable-configuration-package


CHANGES.md
# Changes

## 2.12.0

A change from Thomas Junghans now [allows to have sharable configs][pull-24].
It works similar to sharable configs in eslint. Check out the readme for
instructions.

- [`1d5b9a2`](https://github.com/mantoni/min-webdriver/commit/1d5b9a23bb7b2effef1938b8b98a5538f87b2971)
  feature: add shareable config support (Thomas Junghans)
- [`a3ce836`](https://github.com/mantoni/min-webdriver/commit/a3ce836fd9550b0171f75a5ca2297de280d6649c)
  Add contributor
- [`df66b9d`](https://github.com/mantoni/min-webdriver/commit/df66b9d85e314c50abd91721833dfc8cfc82a2ca)
  Update Studio Changes for `--footer` support

_Released by [Maximilian Antoni](https://github.com/mantoni) on 2018-09-05._

[pull-24]: https://github.com/mantoni/min-webdriver/pull/24

## 2.11.2

- [`7728d9f`](https://github.com/mantoni/min-webdriver/commit/7728d9f5c183e28b173af0c69c128dcec968925f)
  Fix issue in request module invoking a callback without both an error and a response (Davide Callegari)
- [`437bc7f`](https://github.com/mantoni/min-webdriver/commit/437bc7f4d07a919028b05428c781f131f7d2bf0e)
  Add contributor (Maximilian Antoni)
- [`25cf173`](https://github.com/mantoni/min-webdriver/commit/25cf173096551a3e60a685c645d4da5612d69b61)
  Update Studio Changes (Maximilian Antoni)
- [`aa3ea84`](https://github.com/mantoni/min-webdriver/commit/aa3ea8476b73813d35f2e12f569257dfc781b5ac)
  Run build in node 10 (Maximilian Antoni)

## 2.11.1

- [`d9eb5d4`](https://github.com/mantoni/min-webdriver/commit/d9eb5d4cc18eb8e4b113e3a4f19a8661138d917b)
  Improve browser title
- [`7fb8c3c`](https://github.com/mantoni/min-webdriver/commit/7fb8c3c0b5e50c6072e7df8abe397b9b9ddd19e9)
  Add commit links with `--commits`
- [`32a3e02`](https://github.com/mantoni/min-webdriver/commit/32a3e02094b1c330b07b72eb76bbe6a91ffa2468)
  Replace jslint with eslint
- [`66bd668`](https://github.com/mantoni/min-webdriver/commit/66bd66853054f1b5f29c5b95c9862b09c0ec4402)
  Update dependencies

## 2.11.0

Steffen André Langnes added the ability to configure custom capabilities.

- Allow custom capabilities
- Fix example in readme for loading .min-wd file as module
- Add BrowserStack example to readme

## 2.10.0

- Support Appium with iOS Simulator
- Fix "SauceLabs on Travis" link
- Update Mocha link (Morton Fox)

## 2.9.3

Decrease buffer size even further due to Chrome issue still occurring.

## 2.9.2

Further decreases the buffer size introduced in `v2.9.1` to work around the
Chrome driver issue which still occurs occasionally.

## 2.9.1

Running test suits or projects sometimes caused issues with Chrome. This appears
to be an issue with a buffer size limit when executing scripts through
`chromedriver`.

With this path, scripts are no longer injected directly. Instead, a small
receiver function is injected and the actual test code is sent to that function
in chunks, bypassing the buffer limit. Once the script was fully received by the
browser, it is injected with a new `<script>` tag.

Related issues:

- <https://github.com/mantoni/mochify.js/issues/110>
- <https://github.com/sinonjs/sinon/issues/912>
- <https://bugs.chromium.org/p/chromedriver/issues/detail?id=402>

## 2.9.0

Moshe Kolodnya [made it possible][PR14] to use a different config file. He also
[sent a PR][PR13] to allow the `.min-wd` file to be loaded as a node module.
Thanks!

[PR14]: https://github.com/mantoni/min-webdriver/pull/14
[PR13]: https://github.com/mantoni/min-webdriver/pull/13

## 2.8.0

- Streams 3: bump dependencies

## 2.7.0

- Add the option to test using synchronous polling (Matheus Kautzmann)
- Bump saucelabs version to `^1.0` (Matheus Kautzmann)

## 2.6.1

- Update through2 to `^1.1`

## 2.6.0

- Default `sauceJobName` to the name property in cwd `package.json`

## 2.5.1

- Improve documentation for SauceLabs specific options

## 2.5.0

- Update Sauce Job with passed and build attribute (az7arul)

## 2.4.2

- Fix `write after end` bug

## 2.4.1

- Start log polling more quickly
- Write out parsed response status code in failure cases

## 2.4.0

- Pipe output back to browserify instead of stdout

## 2.3.0

- Add `closeOnError` and `closeOnSuccess` options (Dustin Wyatt)

## 2.2.2

- Print browser names with leading '#'

## 2.2.1

- Wait for queue drain before continue polling
- Don't end queue twice

## 2.2.0

- Bump resolve and source-mapper to ^1.0

## 2.1.1

- Log error output to `context.out` instead of `process.stderr`

## 2.1.0

- Map stack traces to original sources using inline source maps

## 2.0.2

- Pipe output to stdout again

## 2.0.1

- Remove brout again

## 2.0.0

- Convert from transform into plugin

## 1.0.0

- Use brout 1.0

## 0.4.5

- Only inject `require` into `.js` files
- Attempt to parse error message from Selenium WebDriver server

## 0.4.4

- Bump brout

## 0.4.3

- Set timeout to `0` to skip setting a log polling timeout

## 0.4.2

- Load a url for a specific browser (for IE 9 doctype workaround)
- Configure timeout for log polling
- Improved log flushing
- Improved http error handling for 302 / 303 responses

## 0.4.1

- Stream browser output instead of buffering per browser. Prints logs from
  first browser in the queue as they arrive, buffering up the other ones.

## 0.4.0

- SauceLabs support
- Load options from package.json

## 0.3.0

- Support for custom start URL

## 0.2.0

- Added callback API
- Simplified transform

## 0.1.0

- Initial release


test/options-test.js
'use strict';
/*eslint-env mocha*/
var path = require('path');
var fs = require('fs');
var assert = require('assert');
var sinon = require('sinon');
var options = require('../lib/options');

describe('options', function () {
  var sandbox = sinon.createSandbox();

  this.afterEach(function () {
    sandbox.restore();
  });

  it('returns defaults', function () {
    assert.deepEqual(options.retrieve({}), {
      hostname: 'localhost',
      port: 4444,
      browsers: [{ name: 'chrome' }],
      asyncPolling: true,
      pollingInterval: 1000,
      closeOnError: true,
      closeOnSuccess: true
    });
  });

  describe('_loadOptions', function () {
    it('returns false if specified file does not exist', function () {
      sandbox.stub(fs, 'existsSync').returns(false);

      assert.equal(options._loadOptions(), false);
    });

    it('loads overrides from js file', function () {
      var fileName = path.resolve(__dirname, 'fixtures', 'options.js');
      var opts = {
        hostname: 'localhost',
        port: 4444
      };
      assert.ok(options._loadOptions(fileName, null, opts));
      assert.deepEqual(opts, {
        hostname: 'my.local',
        port: 4444
      });
    });

    it('loads overrides from text file', function () {
      var fileName = path.resolve(__dirname, 'fixtures', '.min-wd');
      var opts = {
        hostname: 'localhost',
        port: 4444
      };
      assert.ok(options._loadOptions(fileName, null, opts));
      assert.deepEqual(opts, {
        hostname: 'my.local',
        port: 4444
      });
    });
  });

  describe('shareable config override', function () {
    var defaultCwd = process.cwd();

    afterEach(function () {
      process.chdir(defaultCwd);
    });

    it('applies config specified in package.json', function () {
      process.chdir(path.resolve(__dirname, 'fixtures', 'component-a'));

      assert.equal(options.retrieve({}).hostname, 'my.local');
    });

    it('throws if wrong config name is given to `extends`', function () {
      process.chdir(path.resolve(__dirname, 'fixtures', 'component-b'));

      assert.throws(function () {
        options.retrieve({});
      });
    });

    it('ignores missing `webdriver.extends` in config package.json',
      function () {
        process.chdir(path.resolve(__dirname, 'fixtures', 'component-c'));

        assert.equal(options.retrieve({}).hostname, 'localhost');
      });

    it('applies .min-wd config over package.json in shareable config',
      function () {
        process.chdir(path.resolve(__dirname, 'fixtures', 'component-d'));

        assert.equal(options.retrieve({}).hostname, 'min-wd.local');
        assert.equal(options.retrieve({}).port, 2727);
      });

    it('is overridden by .min-wd', function () {
      process.chdir(path.resolve(__dirname, 'fixtures', 'component-d'));

      assert.equal(options.retrieve({}).hostname, 'min-wd.local');
    });
  });
});


test/fixtures/options.js
'use strict';

module.exports = {
  hostname: 'my.local',
};

lib/client.js
/*
 * min-webdriver
 *
 * Copyright (c) 2014 Maximilian Antoni <mail@maxantoni.de>
 *
 * @license MIT
 */
/*global window*/
'use strict';

var brout   = require('brout');
var logs    = [];
var pending = null;
var flushTimeout = null;

function flush() {
  clearTimeout(flushTimeout);
  pending(logs.join(''));
  logs.length = 0;
  pending = null;
}

function push(str) {
  logs.push(str);
  if (pending) {
    flush();
  }
}

window._webdriver_poll = function (callback) {
  flushTimeout = setTimeout(flush, 5000);
  pending = callback;
  if (logs.length) {
    flush();
  }
};

window._webdriver_manualPoll = function () {
  var str = logs.join('');
  logs = [];
  return str;
};

brout.on('out', push);
brout.on('err', push);
brout.on('exit', function (code) {
  push('\nWEBDRIVER_EXIT(' + code + ')\n');
});


lib/plugin.js
/*
 * min-webdriver
 *
 * Copyright (c) 2014-2015 Maximilian Antoni <mail@maxantoni.de>
 *
 * @license MIT
 */
'use strict';

var path    = require('path');
var through = require('through2');
var driver  = require('../lib/driver');
var options = require('../lib/options');


function relative(file) {
  return './' + path.relative(process.cwd(), file).replace(/\\/g, '/');
}

module.exports = function (b, opts) {

  var clientFile = path.join(__dirname, 'client.js');
  b.add(relative(clientFile), {
    expose : 'min-wd'
  });

  var done;
  var stream = through();
  var wrap = b.pipeline.get('wrap');
  wrap.push(through(function (chunk, enc, next) {
    stream.push(chunk);
    next();
  }, function (next) {
    done = next;
    stream.push(null);
  }));
  wrap.push(driver(stream, options.retrieve(opts || {}), function (err) {
    if (err) {
      b.emit('error', err);
    }
    if (done) {
      done();
    }
  }));

};


lib/options.js
/*
 * min-webdriver
 *
 * Copyright (c) 2014 Maximilian Antoni <mail@maxantoni.de>
 *
 * @license MIT
 */
'use strict';

var fs = require('fs');
var path = require('path');
var resolve = require('resolve');

function override(options, overrides) {
  var k;
  for (k in overrides) {
    if (overrides.hasOwnProperty(k)) {
      options[k] = overrides[k];
    }
  }
}

function loadOptions(fileName, key, options) {
  var overrides;

  if (fs.existsSync(fileName)) {
    try {
      overrides = require(path.resolve(fileName));
    } catch (e) {
      overrides = JSON.parse(fs.readFileSync(fileName, 'utf8'));
    }
    if (key) {
      overrides = overrides[key];
      if (!overrides) {
        return false;
      }
    }
    if (overrides.extends) {
      var shareableConfigPath = path.dirname(resolve
        .sync('min-wd-config-' + overrides.extends, {
          basedir: process.cwd()
        }));

      if (!loadOptions(path.resolve(shareableConfigPath, '.min-wd'), null,
        options)) {
        overrides = loadOptions(path.resolve(shareableConfigPath,
          'package.json'), 'webdriver', options);
      }
    }

    override(options, overrides);
    return true;
  }
  return false;
}

exports.retrieve = function retrieve(opts) {
  var options = {
    hostname      : 'localhost',
    port          : 4444,
    browsers      : [{
      name        : 'chrome'
    }],
    asyncPolling : true,
    pollingInterval : 1000,
    closeOnError  : true,
    closeOnSuccess : true
  };

  if (!loadOptions(opts.wdFile || '.min-wd', null, options)) {
    loadOptions('package.json', 'webdriver', options);
  }

  if (opts) {
    override(options, opts);
  }

  if (options.sauceLabs) {
    options.hostname = 'ondemand.saucelabs.com';
    options.port = 80;
    if (!options.sauceJobName && fs.existsSync('package.json')) {
      var packageJson = JSON.parse(fs.readFileSync('package.json', 'utf8'));
      options.sauceJobName = packageJson.name;
    }
  }

  return options;
};

// export for testing
module.exports._loadOptions = loadOptions;

lib/request.js
/*
 * min-webdriver
 *
 * Copyright (c) 2014 Maximilian Antoni <mail@maxantoni.de>
 *
 * @license MIT
 */
'use strict';

var http = require('http');

var contentType = 'application/json;charset=utf-8';


function indent(string) {
  return string.replace(/^/gm, '    ');
}

function fail(context, options, message, headers, body, callback) {
  context.out.write(options.method);
  context.out.write(' ');
  context.out.write(options.path);
  context.out.write('\n\n');
  context.out.write(message);
  if (headers) {
    context.out.write('\n');
    Object.keys(headers).forEach(function (header) {
      context.out.write('\n' + header + ': ' + headers[header]);
    });
    context.out.write('\n');
  }
  if (body) {
    context.out.write('\n');
    try {
      var parsed = JSON.parse(body);
      context.out.write('Response Status Code: ' + parsed.status
        + '\nResponse Message:\n');
      context.out.write(indent(parsed.value.message) + '\n');
    } catch (e) {
      context.out.write(indent(body.replace(/^\s+$/gm, '')));
    }
    context.out.write('\n');
  }
  context.out.write('\n');
  callback(new Error(message));
}

function request(context, method, path, json, callback) {
  var payload = json ? JSON.stringify(json) : null;
  var options = {
    hostname           : context.hostname,
    port               : context.port,
    path               : context.basePath + path,
    method             : method,
    headers            : {
      Connection       : 'keep-alive',
      Accept           : contentType,
      'Content-Type'   : contentType,
      'Content-Length' : payload ? Buffer.byteLength(payload) : 0
    }
  };

  var req = http.request(options, function (res) {

    var body = '';
    res.setEncoding('utf8');
    res.on('data', function (chunk) {
      body += chunk;
    });
    res.on('end', function () {
      if ((res.statusCode === 302 || res.statusCode === 303)
          && path === '/session') {
        var sessionId = res.headers.location;
        if (!sessionId) {
          fail(context, options, 'Received HTTP status ' + res.statusCode
              + ' without location header', res.headers, body, callback);
          return;
        }
        sessionId = sessionId.substring(sessionId.lastIndexOf('/') + 1);
        callback(null, {
          sessionId : sessionId
        });
        return;
      }
      if (res.statusCode < 200 || res.statusCode >= 300) {
        fail(context, options, 'Unexpected HTTP status: ' + res.statusCode
             + ' ' + http.STATUS_CODES[res.statusCode], res.headers, body,
        callback);
        return;
      }
      if (body) {
        var parsed;
        try {
          parsed = JSON.parse(body);
        } catch (e) {
          fail(context, options, 'Unexpected response:', null, body, callback);
          return;
        }
        if (parsed.status) {
          fail(context, options, 'Unexpected response status: '
               + parsed.status, null, parsed.value.message, callback);
        } else {
          callback(null, parsed);
        }
      } else {
        fail(context, options, 'Unexpected empty body:', null, body, callback);
      }
    });

  });

  req.on('error', function (e) {
    fail(context, options, 'Unexcected request error: ' + e.message, null,
      e.stack, callback);
  });

  if (payload) {
    req.write(payload);
  }
  req.end();
}

module.exports = request;


lib/driver.js
/*
 * min-webdriver
 *
 * Copyright (c) 2014 Maximilian Antoni <mail@maxantoni.de>
 *
 * @license MIT
 */
'use strict';

var through   = require('through2');
var listen    = require('listen');
var sourceMap = require('source-mapper');
var request   = require('./request');
var sauceLabs = require('./saucelabs');


function close(context, callback) {
  request(context, 'DELETE', '/window', null, function () {
    request(context, 'DELETE', '', null, function () {
      callback();
    });
  });
}

function handleError(context, err, callback) {
  if (context.closeOnError) {
    close(context, function () {
      callback(err);
    });
  } else {
    callback(err);
  }
}

function pollLogs(context, callback) {
  var endpoint = context.asyncPolling ? '/execute_async' : '/execute';
  var timeout = context.asyncPolling ? 0 : context.pollingInterval;
  var script;
  if (context.asyncPolling) {
    script = 'window._webdriver_poll(arguments[0]);';
  } else {
    script = 'return window._webdriver_manualPoll();';
  }
  request(context, 'POST', endpoint, {
    script : script,
    args   : []
  }, function (err, res) {
    if (err) {
      handleError(context, err, callback);
      return;
    }
    var match = res.value.match(/^WEBDRIVER_EXIT\(([0-9]+)\)$/m);
    if (match) {
      context.out.write(res.value.substring(0, match.index));

      var localCallback = function () {
        var passed = match[1] === '0';

        function done() {
          if (passed) {
            callback();
          } else {
            callback(new Error('Build failed: ' + match[1]));
          }
        }

        if (context.sauceLabs) {
          sauceLabs.updateJob(res.sessionId, {
            name   : context.sauceJobName,
            passed : passed,
            build  : process.env[context.BUILD_VAR]
          }, done);
        } else {
          done();
        }
      };

      if (context.closeOnSuccess) {
        close(context, localCallback);
      } else {
        localCallback();
      }
      return;
    }
    if (context.out.write(res.value)) {
      setTimeout(function () {
        pollLogs(context, callback);
      }, timeout);
    } else {
      context.out.once('drain', function () {
        setTimeout(function () {
          pollLogs(context, callback);
        }, timeout);
      });
    }
  });
}

/*
 * This hack works around the following issues:
 *
 * https://github.com/mantoni/mochify.js/issues/110
 * https://bugs.chromium.org/p/chromedriver/issues/detail?id=402
 * https://github.com/sinonjs/sinon/issues/912
 *
 * Apparently the Chrome webdriver has a buffer limit somewhere around 1 MB.
 * Injecting scripts that are below a certain size works reliably, so we have
 * to slice the actual script into chunks, merge the parts in the browser and
 * then inject a script tag there.
 */
var MAX_SCRIPT_CHUNK = 700 * 1000;
function streamChunk(context, script, callback) {
  var execute = false;
  var nextScript = '';
  if (script.length > MAX_SCRIPT_CHUNK) {
    nextScript = script.substring(MAX_SCRIPT_CHUNK);
    script = script.substring(0, MAX_SCRIPT_CHUNK);
  } else {
    execute = true;
  }
  request(context, 'POST', '/execute', {
    script: 'window._webdriver_receive(arguments[0], arguments[1])',
    args: [script, execute]
  }, function (err) {
    if (err) {
      handleError(context, err, callback);
    } else if (nextScript) {
      streamChunk(context, nextScript, callback);
    } else {
      setTimeout(function () {
        pollLogs(context, callback);
      }, 10);
    }
  });
}

function execute(context, script, callback) {
  request(context, 'POST', '/execute', {
    script: 'var script = "";'
      + 'window._webdriver_receive = function (chunk, execute) {'
      + 'script += chunk;'
      + 'if (execute) {'
      + '  var s = document.createElement("script");'
      + '  s.textContent = script;'
      + '  document.body.appendChild(s);'
      + '}};',
    args: []
  }, function (err) {
    if (err) {
      handleError(context, err, callback);
    } else {
      streamChunk(context, script, callback);
    }
  });
}

function openUrl(context, script, callback) {
  var browser = context.browser;
  var parts = [browser.name];
  if (browser.version) {
    parts.push(browser.version);
  }
  if (browser.platformName) {
    parts.push(browser.platformName);
  }
  if (browser.platformVersion) {
    parts.push(browser.platformVersion);
  }
  var title = parts.join(' ');
  context.out.write('# ' + title + ':\n');

  var x = sourceMap.extract(script);
  if (x.map) {
    var sm = sourceMap.stream(x.map);
    sm.pipe(context.out);
    context.out = sm;
  }

  var url = browser.url || context.url;
  if (url) {
    request(context, 'POST', '/url', {
      url : url
    }, function (err) {
      if (err) {
        handleError(context, err, callback);
        return;
      }
      execute(context, x.js, callback);
    });
  } else {
    execute(context, x.js, callback);
  }
}

var optional_caps = [
  'platformName',
  'platformVersion',
  'deviceName'
];

function connectBrowser(context, callback) {
  var caps = {
    browserName       : context.browser.name,
    version           : context.browser.version,
    platform          : context.browser.platform,
    javascriptEnabled : true
  };
  if (context.sauceLabs) {
    caps.username  = process.env.SAUCE_USERNAME;
    caps.accessKey = process.env.SAUCE_ACCESS_KEY;
  }
  optional_caps.forEach(function (key) {
    if (context.browser[key]) {
      caps[key] = context.browser[key];
    }
  });
  if (context.browser.capabilities) {
    Object.keys(context.browser.capabilities).forEach(function (key) {
      caps[key] = context.browser.capabilities[key];
    });
  }
  var json = {
    desiredCapabilities : caps
  };

  request(context, 'POST', '/session', json, function (err, res) {
    if (err) {
      callback(err);
      return;
    }
    context.basePath = context.basePath + '/session/' + res.sessionId;
    if (!context.asyncPolling || context.timeout === 0) {
      callback(null);
      return;
    }
    request(context, 'POST', '/timeouts/async_script', {
      ms : context.timeout || 10000
    }, function (err) {
      if (err) {
        handleError(context, err, callback);
      } else {
        callback(null);
      }
    });
  });
}

function createContext(options, browser, out) {
  return {
    hostname      : options.hostname,
    port          : options.port,
    url           : options.url,
    asyncPolling  : options.asyncPolling,
    pollingInterval : options.pollingInterval,
    timeout       : options.timeout,
    basePath      : '/wd/hub',
    browser       : browser,
    BUILD_VAR     : options.BUILD_VAR,
    sauceJobName  : options.sauceJobName,
    out           : out,
    sauceLabs     : options.sauceLabs,
    closeOnError  : options.closeOnError,
    closeOnSuccess : options.closeOnSuccess
  };
}

function pipe(streams, out) {
  if (!streams.length) {
    out.end();
    return;
  }
  var stream = streams.shift();
  stream.on('data', function (data) {
    out.write(data);
  });
  stream.on('end', function () {
    pipe(streams, out);
  });
  stream.resume();
}

function run(options, out, runner, callback) {
  var listener = listen();
  var streams  = options.browsers.map(function (browser) {
    var stream = through();
    stream.pause();
    var cb = listener(function () {
      stream.end();
    });
    var context = createContext(options, browser, stream);
    connectBrowser(context, function (err) {
      if (err) {
        cb(err);
        return;
      }
      runner(context, cb);
    });
    return stream;
  });
  pipe(streams, out);
  listener.then(callback);
}

function createRunner(input) {
  var requests = [];
  var script   = '';
  input.on('data', function (chunk) {
    script += chunk;
  });
  input.on('end', function () {
    if (!script) {
      requests.forEach(function (request) {
        close(request.context, request.callback);
      });
    } else {
      requests.forEach(function (request) {
        openUrl(request.context, script, request.callback);
      });
    }
    requests = null;
  });

  return function (context, callback) {
    if (requests) {
      requests.push({ context : context, callback : callback });
    } else {
      if (!script) {
        close(context, callback);
        return;
      }
      openUrl(context, script, callback);
    }
  };
}


module.exports = function (input, options, callback) {
  var error;
  var out = through(function (chunk, enc, next) {
    this.push(chunk);
    next();
  }, function (next) {
    next();
    callback(error);
  });

  run(options, out, createRunner(input), function (err) {
    error = err;
  });

  return out;
};
